lib/vfscore/main.c

```diff
@@ -1164,39 +1179,93 @@ struct dirent *readdir(DIR *dir)
 
 int readdir_r(DIR *dir, struct dirent *entry, struct dirent **result)
 {
-	int error;
+	/**
+	 * `sys_readdir()` will end up calling `VOP_READDIR`, which expects
+	 * a dirent64 structure instead of the `struct dirent` entry parameter.
+	 */
+	struct dirent64 entry64;
 	struct vfscore_file *fp;
+	int error;
 
 	trace_vfs_readdir(dir->fd, entry);
 	error = fget(dir->fd, &fp);
-	if (!error) {
-		error = sys_readdir(fp, entry);
-		fdrop(fp);
-		if (error) {
-			trace_vfs_readdir_err(error);
-		} else {
-			trace_vfs_readdir_ret();
-		}
+	if (unlikely(error))
+		goto err_out;
+
+	error = sys_readdir(fp, &entry64);
+	fdrop(fp);
+	if (unlikely(error)) {
```

Comment:
Do we really need three `unlikely`s for same `error` :thinking:? And those `else` statements look somewhat unnecessary on the surface :thinking:. What is your opinion on this alternative:
```C
int readdir_r(DIR *dir, struct dirent *entry, struct dirent **result)
{
	/**
	 * `sys_readdir()` will end up calling `VOP_READDIR`, which expects
	 * a dirent64 structure instead of the `struct dirent` entry parameter.
	 */
	struct dirent64 entry64;
	struct vfscore_file *fp;
	int error;

	*result = NULL;

	// Our dirent has (like Linux) a d_reclen field, but a constant size.
	entry->d_reclen = sizeof(*entry);

	trace_vfs_readdir(dir->fd, entry);
	error = fget(dir->fd, &fp);
	if (unlikely(error))
		return error;    // This here can never be `ENOENT`, so goto err_out check is unnecessary

	error = sys_readdir(fp, &entry64);
	fdrop(fp);
	if (unlikely(error)) {
		trace_vfs_readdir_err(error);

		return error == ENOENT ? 0 : error;
	}

	trace_vfs_readdir_ret();

	*result = entry;

	entry->d_ino = entry64.d_ino;
	entry->d_type = entry64.d_type;
	entry->d_off = entry64.d_off;
	memcpy(entry->d_name, entry64.d_name, sizeof(entry64.d_name));

	return 0;
}
```